File: sprint5/By languages/kotlin/lesson10code2.kt
fun pop_max(heap: ArrayList<Int>): Int {
    val result = heap[0]
    heap[0] = heap[heap.size - 1]
    heap.removeAt(heap.size - 1)
    sift_down(heap, 0)
    return result
}

fun sift_down(heap: ArrayList<Int>, index: Int) {
    val left = 2 * index
    val right = 2 * index + 1

    // нет дочерних узлов    
    if (heap.size < left) {
        return
    }

    val index_largest: Int
    // right <= heap.size проверяет, что есть оба дочерних узла
    if (right <= heap.size && heap[left] < heap[right]) {
        index_largest = right
    } else {
        index_largest = left
    }

    if (heap[index] < heap[index_largest]) {
        val temp = heap[index]
        heap[index] = heap[index_largest]
        heap[index_largest] = temp
        sift_down(heap, index_largest)
    }
}

File: sprint5/By lessons/lesson10code2/lesson10code2.cpp
#include <iostream>
#include <vector>

void siftDown(std::vector<int>& heap, int index) {
    int left = 2 * index + 1;
    int right = 2 * index + 2;

    if (left >= heap.size()) {
        return;
    }

    int indexLargest = left;
    if (right < heap.size() && heap[left] < heap[right]) {
        indexLargest = right;
    }

    if (heap[index] < heap[indexLargest]) {
        std::swap(heap[index], heap[indexLargest]);
        siftDown(heap, indexLargest);
    }
}

int popMax(std::vector<int>& heap) {
    int result = heap[0];
    heap[0] = heap.back();
    heap.pop_back();
    siftDown(heap, 0);
    return result;
}

int main() {
    std::vector<int> heap = {8, 5, 3};
    int maxValue = popMax(heap);
    std::cout << maxValue << std::endl; // Output: 8
    for (int value : heap) {
        std::cout << value << " "; // Output: 5 3
    }
    return 0;
}
